payara-micro-4.1.1.162 - [DEP0004:Deployment failed due to the conflict occur for jndi-name]

Getting SEVERE [1] information running payara micro:

``` bash
[2016-07-01T09:55:06.488-0300] [Payara Micro 4.1] [SEVERE] [] [javax.enterprise.system.tools.deployment.dol] [tid: _ThreadID=1 _ThreadName=main] [timeMillis: 1467377706488] [levelValue: 1000] DEP0004:Deployment failed due to the conflict occur for jndi-name: java:app/gace/MyDS for application: gace
```

The DataSource definition is here [2]. Even complaining about the deployment the app works.

[1] https://gist.github.com/gilbertoca/a45b6b74b1e48293c620eb4629d8720b
[2] https://gist.github.com/gilbertoca/f2a386f52373e2826ff7b99926f9f4e7

Changed the name to **GaceDataSource**  and the results 
```bash
[2016-07-11T11:26:06.310-0300] [Payara Micro 4.1] [SEVERE] [] [javax.enterprise.system.core] [tid: _ThreadID=1 _ThreadName=main] [timeMillis: 1468247166310] [levelValue: 1000] Exception while invoking class org.glassfish.persistence.jpa.JPADeployer prepare method

java.lang.RuntimeException: Invalid resource : GaceDataSource__pm
        at com.sun.enterprise.connectors.ConnectorRuntime.lookupDataSourceInDAS(ConnectorRuntime.java:593)
        at com.sun.enterprise.connectors.ConnectorRuntime.lookupPMResource(ConnectorRuntime.java:517)
        at org.glassfish.persistence.common.PersistenceHelper.lookupPMResource(PersistenceHelper.java:63)
        at org.glassfish.persistence.jpa.ProviderContainerContractInfoBase.lookupDataSource(ProviderContainerContractInfoBase.java:71)
        at org.glassfish.persistence.jpa.PersistenceUnitInfoImpl.<init>(PersistenceUnitInfoImpl.java:108)
        at org.glassfish.persistence.jpa.PersistenceUnitLoader.loadPU(PersistenceUnitLoader.java:150)
        at org.glassfish.persistence.jpa.PersistenceUnitLoader.<init>(PersistenceUnitLoader.java:114)
        at org.glassfish.persistence.jpa.JPADeployer$1.visitPUD(JPADeployer.java:223)
        at org.glassfish.persistence.jpa.JPADeployer$PersistenceUnitDescriptorIterator.iteratePUDs(JPADeployer.java:510)
        at org.glassfish.persistence.jpa.JPADeployer.createEMFs(JPADeployer.java:230)
        at org.glassfish.persistence.jpa.JPADeployer.prepare(JPADeployer.java:168)
        at com.sun.enterprise.v3.server.ApplicationLifecycle.prepareModule(ApplicationLifecycle.java:926)
        at com.sun.enterprise.v3.server.ApplicationLifecycle.deploy(ApplicationLifecycle.java:435)
        at com.sun.enterprise.v3.server.ApplicationLifecycle.deploy(ApplicationLifecycle.java:220)
        at org.glassfish.deployment.admin.DeployCommand.execute(DeployCommand.java:487)
        at com.sun.enterprise.v3.admin.CommandRunnerImpl$2$1.run(CommandRunnerImpl.java:539)
        at com.sun.enterprise.v3.admin.CommandRunnerImpl$2$1.run(CommandRunnerImpl.java:535)
        at java.security.AccessController.doPrivileged(Native Method)
        at javax.security.auth.Subject.doAs(Subject.java:360)
        at com.sun.enterprise.v3.admin.CommandRunnerImpl$2.execute(CommandRunnerImpl.java:534)
        at com.sun.enterprise.v3.admin.CommandRunnerImpl$3.run(CommandRunnerImpl.java:565)
        at com.sun.enterprise.v3.admin.CommandRunnerImpl$3.run(CommandRunnerImpl.java:557)
        at java.security.AccessController.doPrivileged(Native Method)
        at javax.security.auth.Subject.doAs(Subject.java:360)
        at com.sun.enterprise.v3.admin.CommandRunnerImpl.doCommand(CommandRunnerImpl.java:556)
        at com.sun.enterprise.v3.admin.CommandRunnerImpl.doCommand(CommandRunnerImpl.java:1464)
        at com.sun.enterprise.v3.admin.CommandRunnerImpl.access$1300(CommandRunnerImpl.java:109)
        at com.sun.enterprise.v3.admin.CommandRunnerImpl$ExecutionContext.execute(CommandRunnerImpl.java:1846)
        at com.sun.enterprise.v3.admin.CommandRunnerImpl$ExecutionContext.execute(CommandRunnerImpl.java:1722)
        at com.sun.enterprise.admin.cli.embeddable.DeployerImpl.deploy(DeployerImpl.java:134)
        at fish.payara.micro.PayaraMicro.deployAll(PayaraMicro.java:1249)
        at fish.payara.micro.PayaraMicro.bootStrap(PayaraMicro.java:905)
        at fish.payara.micro.PayaraMicro.main(PayaraMicro.java:148)
Caused by: com.sun.appserv.connectors.internal.api.ConnectorRuntimeException: Invalid resource : GaceDataSource__pm
        at org.glassfish.jdbcruntime.service.JdbcDataSource.validateResource(JdbcDataSource.java:82)
        at org.glassfish.jdbcruntime.service.JdbcDataSource.setResourceInfo(JdbcDataSource.java:63)
        at org.glassfish.jdbcruntime.JdbcRuntimeExtension.lookupDataSourceInDAS(JdbcRuntimeExtension.java:142)
        at com.sun.enterprise.connectors.ConnectorRuntime.lookupDataSourceInDAS(ConnectorRuntime.java:589)
        ... 32 more
[2016-07-11T11:26:06.313-0300] [Payara Micro 4.1] [SEVERE] [] [javax.enterprise.system.core] [tid: _ThreadID=1 _ThreadName=main] [timeMillis: 1468247166313] [levelValue: 1000] Exception while preparing the app

[2016-07-11T11:26:06.314-0300] [Payara Micro 4.1] [SEVERE] [NCLS-CORE-00026] [javax.enterprise.system.core] [tid: _ThreadID=1 _ThreadName=main] [timeMillis: 1468247166314] [levelValue: 1000] [[
  Exception during lifecycle processing
java.lang.RuntimeException: Invalid resource : GaceDataSource__pm
        at com.sun.enterprise.connectors.ConnectorRuntime.lookupDataSourceInDAS(ConnectorRuntime.java:593)
        at com.sun.enterprise.connectors.ConnectorRuntime.lookupPMResource(ConnectorRuntime.java:517)
        at org.glassfish.persistence.common.PersistenceHelper.lookupPMResource(PersistenceHelper.java:63)
        at org.glassfish.persistence.jpa.ProviderContainerContractInfoBase.lookupDataSource(ProviderContainerContractInfoBase.java:71)
        at org.glassfish.persistence.jpa.PersistenceUnitInfoImpl.<init>(PersistenceUnitInfoImpl.java:108)
        at org.glassfish.persistence.jpa.PersistenceUnitLoader.loadPU(PersistenceUnitLoader.java:150)
        at org.glassfish.persistence.jpa.PersistenceUnitLoader.<init>(PersistenceUnitLoader.java:114)
        at org.glassfish.persistence.jpa.JPADeployer$1.visitPUD(JPADeployer.java:223)
        at org.glassfish.persistence.jpa.JPADeployer$PersistenceUnitDescriptorIterator.iteratePUDs(JPADeployer.java:510)
        at org.glassfish.persistence.jpa.JPADeployer.createEMFs(JPADeployer.java:230)
        at org.glassfish.persistence.jpa.JPADeployer.prepare(JPADeployer.java:168)
        at com.sun.enterprise.v3.server.ApplicationLifecycle.prepareModule(ApplicationLifecycle.java:926)
        at com.sun.enterprise.v3.server.ApplicationLifecycle.deploy(ApplicationLifecycle.java:435)
        at com.sun.enterprise.v3.server.ApplicationLifecycle.deploy(ApplicationLifecycle.java:220)
        at org.glassfish.deployment.admin.DeployCommand.execute(DeployCommand.java:487)
        at com.sun.enterprise.v3.admin.CommandRunnerImpl$2$1.run(CommandRunnerImpl.java:539)
        at com.sun.enterprise.v3.admin.CommandRunnerImpl$2$1.run(CommandRunnerImpl.java:535)
        at java.security.AccessController.doPrivileged(Native Method)
        at javax.security.auth.Subject.doAs(Subject.java:360)
        at com.sun.enterprise.v3.admin.CommandRunnerImpl$2.execute(CommandRunnerImpl.java:534)
        at com.sun.enterprise.v3.admin.CommandRunnerImpl$3.run(CommandRunnerImpl.java:565)
        at com.sun.enterprise.v3.admin.CommandRunnerImpl$3.run(CommandRunnerImpl.java:557)
        at java.security.AccessController.doPrivileged(Native Method)
        at javax.security.auth.Subject.doAs(Subject.java:360)
        at com.sun.enterprise.v3.admin.CommandRunnerImpl.doCommand(CommandRunnerImpl.java:556)
        at com.sun.enterprise.v3.admin.CommandRunnerImpl.doCommand(CommandRunnerImpl.java:1464)
        at com.sun.enterprise.v3.admin.CommandRunnerImpl.access$1300(CommandRunnerImpl.java:109)
        at com.sun.enterprise.v3.admin.CommandRunnerImpl$ExecutionContext.execute(CommandRunnerImpl.java:1846)
        at com.sun.enterprise.v3.admin.CommandRunnerImpl$ExecutionContext.execute(CommandRunnerImpl.java:1722)
        at com.sun.enterprise.admin.cli.embeddable.DeployerImpl.deploy(DeployerImpl.java:134)
        at fish.payara.micro.PayaraMicro.deployAll(PayaraMicro.java:1249)
        at fish.payara.micro.PayaraMicro.bootStrap(PayaraMicro.java:905)
        at fish.payara.micro.PayaraMicro.main(PayaraMicro.java:148)
Caused by: com.sun.appserv.connectors.internal.api.ConnectorRuntimeException: Invalid resource : GaceDataSource__pm
        at org.glassfish.jdbcruntime.service.JdbcDataSource.validateResource(JdbcDataSource.java:82)
        at org.glassfish.jdbcruntime.service.JdbcDataSource.setResourceInfo(JdbcDataSource.java:63)
        at org.glassfish.jdbcruntime.JdbcRuntimeExtension.lookupDataSourceInDAS(JdbcRuntimeExtension.java:142)
        at com.sun.enterprise.connectors.ConnectorRuntime.lookupDataSourceInDAS(ConnectorRuntime.java:589)
        ... 32 more
]]

```
Full log https://gist.github.com/gilbertoca/0d2e707d76399ea81c1191763070b7e7

--- a/appserver/connectors/connectors-runtime/src/main/java/com/sun/enterprise/connectors/service/ConnectorConnectionPoolAdminServiceImpl.java
+++ b/appserver/connectors/connectors-runtime/src/main/java/com/sun/enterprise/connectors/service/ConnectorConnectionPoolAdminServiceImpl.java
@@ -37,7 +37,7 @@
  * only if the new code is made subject to such option by the copyright
  * holder.
  */
-// Portions Copyright [2016] [Payara Foundation]
+// Portions Copyright [2016] [Payara Foundation and/or its affiliates]
 
 package com.sun.enterprise.connectors.service;
 
@@ -188,7 +188,7 @@ public class ConnectorConnectionPoolAdminServiceImpl extends ConnectorService {
         try {
 
             _runtime.getResourceNamingService().publishObject(poolInfo, jndiNameForPool, connectorPoolObj, true);
-            ManagedConnectionFactory mcf = obtainManagedConnectionFactory(poolInfo);
+            ManagedConnectionFactory mcf = createManagedConnectionFactory(connectorPoolObj);
             if (mcf == null) {
                 _runtime.getResourceNamingService().unpublishObject(poolInfo, jndiNameForPool);
                 String i18nMsg = localStrings.getString("ccp_adm.failed_to_create_mcf", poolInfo);
@@ -909,6 +909,13 @@ public class ConnectorConnectionPoolAdminServiceImpl extends ConnectorService {
 
     public ManagedConnectionFactory obtainManagedConnectionFactory(PoolInfo poolInfo) throws ConnectorRuntimeException{
         return obtainManagedConnectionFactory(poolInfo, null);
+    }
+
+    private ManagedConnectionFactory createManagedConnectionFactory(ConnectorConnectionPool connectorPoolObj) throws ConnectorRuntimeException {
+        if(_registry.isMCFCreated(connectorPoolObj.getPoolInfo())) {
+            recreateConnectorConnectionPool(connectorPoolObj);
+        }
+        return obtainManagedConnectionFactory(connectorPoolObj.getPoolInfo());
     }
 
     /**

--- a/appserver/deployment/dol/src/main/java/com/sun/enterprise/deployment/util/ApplicationValidator.java
+++ b/appserver/deployment/dol/src/main/java/com/sun/enterprise/deployment/util/ApplicationValidator.java
@@ -91,6 +91,7 @@ public class ApplicationValidator extends ComponentValidator
     final String JNDI_COMP = "java:comp";
     final String JNDI_MODULE = "java:module";
     final String JNDI_APP = "java:app";
+    private final String EJB_LEVEL = "EJBLevel:";
 
     private boolean allUniqueResource = true;
 
@@ -589,14 +590,15 @@ public class ApplicationValidator extends ComponentValidator
                     }
                 }
 
-                Vector vectorScope = commonResourceValidator.getScope();
+                @SuppressWarnings("unchecked")
+                Vector<String> vectorScope = commonResourceValidator.getScope();
                 if (vectorScope != null) {
                     vectorScope.add(scope);
                 }
                 commonResourceValidator.setScope(vectorScope);
                 allResourceDescriptors.put(name, commonResourceValidator);
             } else {
-                Vector<String> vectorScope = new Vector<String>();
+                Vector<String> vectorScope = new Vector<>();
                 vectorScope.add(scope);
                 allResourceDescriptors.put(name, new CommonResourceValidator(descriptor, name, vectorScope));
             }
@@ -709,9 +711,13 @@ public class ApplicationValidator extends ComponentValidator
                     otherElements = otherElements.substring(0, otherElements.indexOf("#"));
                 }
                 if (firstElement.equals(otherElements)) {
+                    boolean fail = !firstElement.startsWith(EJB_LEVEL);
                     inValidJndiName = jndiName;
-                    DOLUtils.getDefaultLogger().log(Level.SEVERE, "DEP0004:Deployment failed due to the conflict occur for jndi-name: {0} for application: {1}",
+                    DOLUtils.getDefaultLogger().log(fail? Level.SEVERE : Level.FINE, "DEP0004:Deployment failed due to the conflict occur for jndi-name: {0} for application: {1}",
                         new Object[] { jndiName, application.getAppName() });
+                    if(fail) {
+                        return false;
+                    }
                 }
             }
         }

--- a/appserver/jdbc/jdbc-ra/jdbc-core/src/main/java/com/sun/gjc/common/DataSourceSpec.java
+++ b/appserver/jdbc/jdbc-ra/jdbc-core/src/main/java/com/sun/gjc/common/DataSourceSpec.java
@@ -41,6 +41,10 @@
 
 package com.sun.gjc.common;
 
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
 import java.util.concurrent.ConcurrentHashMap;
 
 /**
@@ -107,9 +111,10 @@ public class DataSourceSpec implements java.io.Serializable {
     public static final int MODULENAME = 47;
     public static final int SLOWSQLLOGTHRESHOLD = 48;
     public static final int LOGJDBCCALLS = 49;
-    
 
-    private ConcurrentHashMap<Integer, String> details = new ConcurrentHashMap<Integer, String>();
+    private static final long serialVersionUID = 1L;
+
+    private final ConcurrentHashMap<Integer, String> details = new ConcurrentIgnoredHashMap(URL);
 
     /**
      * Set the property.
@@ -141,6 +146,7 @@ public class DataSourceSpec implements java.io.Serializable {
      *
      * @param obj Instance of <code>DataSourceSpec</code> object.
      */
+    @Override
     public boolean equals(Object obj) {
         if (obj instanceof DataSourceSpec) {
             return this.details.equals(((DataSourceSpec) obj).details);
@@ -153,7 +159,74 @@ public class DataSourceSpec implements java.io.Serializable {
      *
      * @return hashCode of this object.
      */
+    @Override
     public int hashCode() {
         return this.details.hashCode();
+    }
+
+
+    /**
+     * If any keys don't exist from the other side, ignore the differences,
+     * as specified by the ignoredList
+     */
+    private static class ConcurrentIgnoredHashMap extends ConcurrentHashMap<Integer, String>
+    {
+        public ConcurrentIgnoredHashMap(Integer... ignoredList) {
+            this.ignoredList = Arrays.asList(ignoredList);
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if(!(o instanceof ConcurrentIgnoredHashMap)) {
+                return false;
+            }
+            ConcurrentIgnoredHashMap rhs = (ConcurrentIgnoredHashMap)o;
+            if(checkIgnoredKeys(rhs) == Containment.ONE_CONTAINS) {
+                return makeIgnoredMap(this).equals(makeIgnoredMap(rhs));
+            } else {
+                return super.equals(rhs);
+            }
+        }
+
+        @Override
+        public int hashCode() {
+            if(checkIgnoredKeys(null) != Containment.NONE_CONTAIN) {
+                return makeIgnoredMap(this).hashCode();
+            }
+            else {
+                return super.hashCode();
+            }
+        }
+
+        private Map<Integer, String> makeIgnoredMap(ConcurrentIgnoredHashMap map) {
+            Map<Integer, String> rv = new HashMap<>(map);
+            for(Integer ignoredValue : ignoredList) {
+                rv.remove(ignoredValue);
+            }
+            return rv;
+        }
+
+        private Containment checkIgnoredKeys(ConcurrentIgnoredHashMap rhs) {
+            for(Integer ignoredValue : ignoredList) {
+                int numContainment = 0;
+                if(containsKey(ignoredValue)) {
+                    ++numContainment;
+                }
+                if(rhs != null && rhs.containsKey(ignoredValue)) {
+                    ++numContainment;
+                }
+                if(numContainment > 0) {
+                    return numContainment > 1? Containment.BOTH_CONTAIN : Containment.ONE_CONTAINS;
+                }
+            }
+            return Containment.NONE_CONTAIN;
+        }
+
+
+        private enum Containment { NONE_CONTAIN, ONE_CONTAINS, BOTH_CONTAIN };
+
+
+        private final List<Integer> ignoredList;
+        private static final long serialVersionUID = 1L;
     }
 }

